joomla/joomla-cms
# Add afterRespond event to the Joomla Debug Console

#### Description

Following https://github.com/joomla/joomla-cms/pull/8888, this PR adds the final event `afterRespond` to the Joomla! Debug Console `Profile Information` block.

As it is now Joomla is not returning the real page time because it's not counting the `afterRespond` event.

To achive this goal a new event, the `onBeforeEnd`, is created. This event:
- is the last one (load after the `onAfterRespond` events) and before the app closes.
- can only be loaded by the debug system plugin and it was to be enabled.

Since the `onAfterRespond` events are now in the Profiler too, this PR will also allow to debug the `onAfterRespond` events.

Before PR:
![image](https://cloud.githubusercontent.com/assets/9630530/12278208/b02e2976-b977-11e5-8411-e4d4c343c96c.png)

After PR:
![image](https://cloud.githubusercontent.com/assets/9630530/12278187/8e778b74-b977-11e5-9059-28a9fdd8947a.png)
#### How to test
1. Apply this patch.
2. Turn on the debug in global config and enale the debug system plugin.
3. Test any page in frontend or admin and check if the `afterRespond` time and memory usage is in the Joomla! Debug Console `Profile Information` block.
4. You can now also test send info to Joomla! Debug Console on a `onAfterRespond`. For instance, add the following php to a system plugin that is enabled, load page and observe if the messages appear in the Joomla! Debug Console.

``` php
public function onAfterRespond()
{
    // Test to add mark in the profiler Joomla! Debug Console Profile Information block.
    JDEBUG ? JProfiler::getInstance('Application')->mark('testMark') : null;

    // Test to add log message to Log Messages block.
    JDEBUG ? JLog::add('Test log message to Joomla! Debug Console Log Messages block.', JLog::WARNING, 'deprecated') : null;
    }
```
#### Observations

Improvements or suggestions are welcome.
